sw/host/opentitanlib/src/test_utils/bitbanging.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod i2c {
+    use anyhow::{anyhow, bail, Result};
+    use std::iter::Peekable;
+
+    #[derive(Debug, PartialEq)]
+    pub enum Transfer<'w> {
+        Start,
+        Stop,
+        Addr(u8, bool),
+        Write(&'w [u8]),
+        WriteNibble(u8), // Can be used to simulate a broken byte.
+        Read(usize),
+        Byte(u8, bool),
+        Bytes(&'w [u8]),
+    }
+
+    impl Transfer<'_> {
+        fn parse<const SDA: u8, const SCL: u8>(&self, is_next_stop: bool) -> Vec<u8> {
+            match *self {
+                Transfer::Start => vec![
+                    0x01 << SDA | 0x01 << SCL,
+                    0x00 << SDA | 0x01 << SCL,
+                    0x00 << SDA | 0x00 << SCL,
+                ],
+                Transfer::Stop => vec![
+                    0x00 << SDA | 0x01 << SCL,
+                    0x01 << SDA | 0x00 << SCL,
+                    0x01 << SDA | 0x01 << SCL,
+                ],
+                Transfer::Addr(byte, read) => self.byte::<SDA, SCL>((byte << 1) | read as u8, read),
```

Comment:
Should this be
```suggestion
                Transfer::Addr(byte, read) => self.byte::<SDA, SCL>((byte << 1) | read as u8, false),
```
?

I am not following why the ACK bit depends on whether it's read or write.

Reply:
When the host is addressing to write it expects the ACK from the target. Otherwise, the host should ACK. Does that make sense? I didn't read the full spec, this is my understanding from experiments with opentitan i2c.

Reply:
I think it should be the target responding the ACK regardless read/write for the address byte.

Reply:
I found this interpretation of the [spec](https://onlinedocs.microchip.com/pr/GUID-CA6B1F8F-E09D-4780-A52A-CBBF61902464-en-US-2/index.html?GUID-FB69465D-2145-4692-9988-FC0715E78C41).
Basically, the transmitter should leave the sda to the receiver to acknowledge the reception. Therefore, when the Host is writing, the target should ack and when the target is writing, then the host should ack.

Reply:
It sounds like that aligns with my understanding that the bit should be always pulled up when writing the address, regardless whether it's read or write.

Reply:
You are right, fixed.